### ISSUE TITLE
Can't create SAML-based AAD apps using az ad app create

### ISSUE THREAD

**Issue by mattdot** [Azure/azure-cli#7579]
**Is your feature request related to a problem? Please describe.**
I can't create a SAML-based SSO AAD app registration using the az CLI.  

It's possible to create it manually in the Azure portal by going to "Enterprise Applications" > "New Application".  The by choosing "Non-gallery application".  Apps created this way show a "Manage > Single sign-on" configuration blade which is not available in apps created using the az CLI.

From my research this seems to be caused because the CLI does not have an option to set the proper Portal UI tag.  Showing it as an enterprise application in portal requires the Portal UI tag of `WindowsAzureActiveDirectoryIntegratedApp`.

**Describe the solution you'd like**
I'd like a way to create SAML-based SSO applications with the az CLI, perhaps using a flag to differentiate between standard apps and integrated SSO apps. 
```
az ad app create -n "My SSO App" --integrated-sso
```

**Describe alternatives you've considered**
The only alternative is to create this manually which breaks automated deployments and management.  This also does not appear to be possible in Powershell CLI.

**Additional context**
Once the app has been created manually, it can be managed using the current az CLI to perform `az ad app update` and change it's properties, however configuring it to be a SAML app does not appear to be possible.

**Comment by tjprescott** [comment]
@yugangw-msft

**Comment by yugangw-msft** [comment]
Need more swaggers.
@mattdot, are you able to do it through [AzureAD's powershell](https://docs.microsoft.com/en-us/powershell/azure/active-directory/overview?view=azureadps-2.0)?

**Comment by b-b3rn4rd** [comment]
It's really disappointing that Azure does not believe in automation and expects engineers to provision resources manually. Especially, when you already have the functionality to do it using UI why wouldn't you put a little bit of extra effort and provide CLI support...

**Comment by yugangw-msft** [comment]
@b-b3rn4rd, Azure CLI's graph commands have never meant to be used for general graph object management; rather they are just enough for RBAC where SAML-based SSO applications don't play any specific role there. Hence we usually don't prioritize such asks till we have enough users votes.

**Comment by m4r10k** [comment]
@b-b3rn4rd you are fully correct on this topic. @yugangw-msft -> We need to integrate 100s of AWS accounts with AAD too and everything which isn't automatable today, will not be taken into account as enterprise ready tomorrow. In addition, the Go docs are really hard to read as things like the following are not really useful, because there is no information about what URI *strings* are needed or if they are created automatically and even if, what happens if I add an uri (append, overwrite, ...)?
~~~
    // IdentifierUris - A collection of URIs for the application.
    IdentifierUris *[]string `json:"identifierUris,omitempty"`
~~~

But the really bad thing is, that there is no interface for the claims, not even a read only interface. I cannot show the claims which I've provided, nor can I configure them via the SDK. :sick:

According to https://docs.microsoft.com/en-gb/azure/active-directory/saas-apps/amazon-web-service-tutorial one has to specify the claim providers manually every time you create an app, even if you choose "AWS" from the gallery app. Jesus, this claims are *always* the same ones! :see_no_evil:

That's stupid and error prone work, that's the key why mankind has invented automation - computers are better at it, if we would have some API interfaces.

@rauschbit

**Comment by yugangw-msft** [comment]
Folks, I will take a look and get back to you.

**Comment by yugangw-msft** [comment]
Folks, I ended up updating the [Rest spec](https://github.com/Azure/azure-rest-api-specs/pull/5181) to open up the support for this. With that, the command to address this issue would be like below. Please let me know if I missed anything. 
It is different from what was proposed by @mattdot, but let us focus on the API readiness before polish  further. BTW, I have a private installer (windows msi, and docker image) for initial test. If you like to try out, I can be reached at yugangw at microsoft dot com 
```bash
   az ad sp create-for-rbac -n sp-reload --skip-assignment
   {
     "appId": "b8f8f0ab-2885-4bb2-9b03-1a514a2bbdc1",
     "displayName": "my-sp",
     "name": "http://my-sp",
     "password": "28999ddb-5ebb-47dd-99ac-a9084215f707",
     "tenant": "54826b22-38d6-4fb2-bad9-123456788888"
   } 

   az ad sp update --id b8f8f0ab-2885-4bb2-9b03-1a514a2bbdc1 --add tags WindowsAzureActiveDirectoryIntegratedApp

   az ad app update --id b8f8f0ab-2885-4bb2-9b03-1a514a2bbdc1 --set samlMetadataUrl=https://mysaml
```

**Comment by m4r10k** [comment]
@yugangw-msft Wow! Thank you for your fast response and improvements 💪 👏! I will contact you via email later for the docker image etc. to try it out!

**Comment by m4r10k** [comment]
@yugangw-msft we have tried it now, but there seams to be still a gap. If I am issuing the provided commands, I got an enterprise application but the SSO option(s) are still missing there.

![grafik](https://user-images.githubusercontent.com/9364572/53401570-5a171400-39b0-11e9-851b-0add7400ff46.png)

The commands itself are clear but how should we proceed from here? I thing there are two steps to take into account:
- First: Someone is starting by choosing a gallery app, for example Amazon Web Services (AWS), which is then configured in detail

![grafik](https://user-images.githubusercontent.com/9364572/53402113-62bc1a00-39b1-11e9-8b1a-e2ed10efb9ef.png)

- Second: Someone is starting by creating a "own app" and is configurating it after the creation

![grafik](https://user-images.githubusercontent.com/9364572/53402215-9dbe4d80-39b1-11e9-87a9-57651cdf4b93.png)

From my point of view it would make sense to have an API or to open up the API for the CLI/SDK which is controlling the screenshots above, "own app" and "gallery app" on the `az ad app create` command for example.

And then I think there would be the need to get in touch with the API's that are behind this stuff:
![grafik](https://user-images.githubusercontent.com/9364572/53402802-b0855200-39b2-11e9-8b36-9ebe2f5a9a06.png)

This would be, probably some kind of `az ad app update....` and yes I am at your side, there is the need, that the REST API's for this must be opened for the CLI/SDK's.

I hope this is helpful...

**Comment by b-b3rn4rd** [comment]
Until this is fully supported, I came up with a simple automation process of pushing IAM roles back to Azure from across all AWS accounts in an organization. In case anyone is interested https://github.com/b-b3rn4rd/aws-saml-azuread

**Comment by yugangw-msft** [comment]
Folks, I would appreciate if someone can send me the application/service principal details using "az ad app show" and "az ap sp show". Through the diff with the regular object, I can update the Rest API to get this supported. I can try all things out myself, but that would take longer time to get to the solution.
Again, I can be reached at yugangw at microsoft dot com

**Comment by m4r10k** [comment]
Sure! I will send it to you later.

**Comment by annbrady** [comment]
any recent updates?  I'm looking to do the same things for the same reasons.  thanks!

**Comment by m4r10k** [comment]
@annbrady I am still in contact with @yugangw-msft via e-mail - offline to this issue. He told me, that it will take until the end of April approximately that the backend rest interfaces will be opened and the SDK's will be able to access it. We have to wait a little bit, at this point, but I am happy that @yugangw-msft is working on it together with the other Azure teams. 😄

**Comment by yugangw-msft** [comment]
Hi folks, we are actively working on this, but at this moment we are not able to offer any ETAs you should align/plan your work with. One thing for sure is once the Service API goes public, CLI team will prioritize the work to expose the support as early as possible.

**Comment by avanbecelaere** [comment]
I'm currently waiting on this as well. Is there an updated status?

**Comment by hemant6488** [comment]
Any ETA on this yet?

**Comment by m4r10k** [comment]
I have no new information on this, but I can ask back.

**Comment by viresh-contino** [comment]
Hi, i am very keen to see this bit of automation!  Has anybody found a workaround ?

**Comment by m4r10k** [comment]
There's currently a work in progress going on behind the scenes. I've asked back and currently there is some kine of alpha work in progress -> https://docs.microsoft.com/en-us/graph/api/resources/applicationtemplate?view=graph-rest-beta to get application templates accessible via the API. But this will still take some time.

There is no workaround available beside doing it manually because the API is simply missing for automation. 😄

**Comment by StarkCaptain** [comment]
As a workaround you can use the graph api beta endpoint to create a SAML application based on the standard SAML application template ID. This will create a base SAML application in Azure AD that you can then update the SAML metadata from. 

https://docs.microsoft.com/en-us/graph/api/applicationtemplate-instantiate?view=graph-rest-beta&tabs=http

The ID of the basic SAML application template from Microsoft is: 8adf8e6e-67b2-4cf2-a259-e3dc5476c621

The endpoint URI would be the below then for creating the application with a request body json object of displayName, like below

```
Request Type: Post

Request Body:
{"displayName":"My App Test"}

URI Endpoint:
https://graph.microsoft.com/beta/applicationTemplates/8adf8e6e-67b2-4cf2-a259-e3dc5476c621/instantiate
```

You could also specify the AWS template ID, which is 8b1025e4-1dd2-430b-a150-2ef79cd700f5

You can search for other template IDs from the graph API endpoint such as below, using displayName as a filter

`https://graph.microsoft.com/beta/applicationTemplates?$filter=contains(displayName, 'AWS')`

or list all of them 

`https://graph.microsoft.com/beta/applicationTemplates
`

**Comment by duduz** [comment]
Is there any progress on this?

same as @m4r10k , i have like 200 AWS accounts i need to integrate with AAD.

**Comment by jiasli** [comment]
@duduz, as AAD is deprecating AD Graph and migrating to MS Graph, Azure CLI is also planning on moving to MS Graph. Temporarily, could you try the solution provided by @StarkCaptain and see if it works for you?

**Comment by duduz** [comment]
@jiasli, i will check it, thank you

**Comment by duduz** [comment]
@jiasli, it does provision the app but a part of the the icon for the app, nothing really being inherited to the new provisioned application
when provisioning from applicationTemplates, ending up with servicePrincipal and an application objects
application doesn't inherit an randomly generated identifierUri
servicePrincipal doesn't inherit the additional claims required, in addition, there's no KeyCredential being generated

all of this, is being provided when provisioning through the UI

When you're doing this through the UI.. everything is good.
something with the API is not fully baked

**Comment by jiasli** [comment]
Hi @duduz, as for the API issue, since it is not under Azure CLI's coverage yet, do you mind creating a support ticket and contact AAD team directly? Thanks for understanding.

**Comment by duduz** [comment]
Hi @jiasli, will do. in addition, i've raised a issue in azure-rest-api-specs repo.

**Comment by Mina69** [comment]
@yugangw-msft Any update on this case?

**Comment by dazinator** [comment]
Any update on this? We currently automate the provisioning of QA environments to azure, however as part of that automation we want the QA engineer to be able to log into the app with their azure ad account via saml. This means we have to automate the creation of a SAML app in azure AD for that QA environment (our qa environments have different URLs). Very disappointing to see this is apparently not yet possible even after all this time?

**Comment by AliKhyar** [comment]
@yugangw-msft any update on this please?

**Comment by jiasli** [comment]
We currently don't have plan to implement SAML-related features in the existing `az ad` module which still uses AD Graph. Please see https://github.com/Azure/azure-cli/issues/12946 for calling Microsoft Graph API directly with `az rest`.

**Comment by jiasli** [comment]
Due to the high number of votes, I am marking this issue as a Feature Candidate, but as stated in https://github.com/Azure/azure-cli/issues/7579#issuecomment-460895548, Azure CLI command module `az ad` is only designed for interacting with Azure RBAC, so SAML-based app is currently not within Azure CLI's support scope.

AAD and Microsoft Graph have their own command line tools:

- Azure AD PowerShell: https://learn.microsoft.com/en-us/powershell/azure/active-directory/overview
- Microsoft Graph PowerShell: https://learn.microsoft.com/en-us/powershell/microsoftgraph/overview
- Microsoft Graph CLI: https://github.com/microsoftgraph/msgraph-cli

These command line tools are better places for supporting SAML-based AAD apps.

**Comment by MaddyMicrosoft** [comment]
The /applicationTemplates/{id}/instantiate endpoint that @StarkCaptain shared as a beta workaround is now GA in v1.0, and Microsoft Graph has a dedicated end-to-end tutorial for this exact scenario:

[Configure SAML-based single sign-on using Microsoft Graph ](https://learn.microsoft.com/en-us/graph/application-saml-sso-configure-api)(uses AWS IAM Identity Center as the walkthrough)

Closing as out of scope for az ad per @jiasli's routing above. If you hit a specific gap against Microsoft Graph, please open a fresh issue via [Microsoft Graph support](https://developer.microsoft.com/graph/support) so it lands with the right team.
